### PR TITLE
Handle explicit recast from base class.

### LINKED ISSUE #45: Recasting from base isn't pretty
Let's say somehow I get an (unsaved) `Animal` instance. There isn't a pretty way to recast as a `Canine` or `Feline`:

```
animal = Animal('phoebo')

# This is a no-op thanks to https://github.com/craigds/django-typed-models/blob/master/typedmodels/models.py#L329
animal.recast(Feline)

# This doesn't work either (unless your model defines __str__ to be <app_name>.<model_name>)
animal.type = Feline

# These will work, but isn't ideal
animal.type = 'my_app.Feline'
animal.recast()

# This works too (but only by coincidence, by how recast is implemented)
animal.type = 'nonesense'
animal.recast(Feline)
```

IMO `animal.recast(Feline)` should be supported.
(setting the `type` is smelly, since `type` is supposed to be a string field)

### PR DESCRIPTION
**Pull request by craigds**
Fixes #45. @thejcannon can you try this branch and see if it solves the problem for you?

### CODE REVIEW

File: typedmodels/models.py [hunk 1]
```diff
@@ -305,20 +305,32 @@ def __init__(self, *args, **kwargs):
...
+        if not self.type:
+            if not hasattr(self, '_typedmodels_type'):
+                # This is an instance of an untyped model
+                if typ is None:
+                    # recast() is probably being called by __init__() here.
+                    # Ideally we'd raise an error here, but the django admin likes to call
+                    # model() and doesn't expect an error.
+                    # Instead, we raise an error when the object is saved.
+                    return
+                else:
```
Comment (thejcannon) [inline typedmodels/models.py]: You don't need this `else` block. As it stands, you'll assign `type`, then on (new) line 339 re-assign `typ`, then on (new) line 346 re-assign `type` from `typ`.

Without the `else` block, (new) line 337 with (new) line 342 handles everything correctly in the `animal.recast(Feline)` use case.

File: typedmodels/tests.py [hunk 2]
```diff
@@ -57,7 +57,14 @@ def animals(db):
     )
 
 
-def test_cant_instantiate_base_model(db):
+def test_can_instantiate_base_model(db):
+    # direct instantiation works fine without a type, as long as you don't save
+    x = Animal()
```
Comment (thejcannon) [inline typedmodels/tests.py]: should this be `animal = ...`?

File: typedmodels/models.py (hunk 1 above)
Comment (craigds) [reply typedmodels/models.py]: ah, good point! I'll nuke this whole chunk 😄

File: typedmodels/models.py [hunk 3]
```diff
@@ -13,7 +13,7 @@
 from django.db.models.fields import Field
 from django.db.models.options import make_immutable_fields_list
 from django.utils.encoding import smart_text
-from django.utils.six import with_metaclass
+from django.utils.six import with_metaclass, string_types
```
Comment (thejcannon) [inline typedmodels/models.py]: You're not using `string_types` anymore

**Comment by thejcannon** [comment]
Aside from unused import, looks great to me. Thanks for getting this done so quickly!

**Pull request merged**